File: python/gen_token.py
APP_VERSION = '2.0.7'

import hashlib
import hmac
import optparse
import sys
import time


class FastlyTokenError(Exception):
    def __init__(self, text):
        self._text = text

    def __str__(self):
        return ':%s' % self._text

    def _getText(self):
        return str(self)
    text = property(_getText, None, None,
        'Formatted error text.')

class FastlyTokenConfig:
    def __init__(self):
        self.ip = ''
        self.start_time = None
        self.end_time = None
        self.lifetime = 300
        self.secret = 'aabbccddeeff00112233445566778899'
        self.stream_id = ''

class FastlyToken:
    def __init__(self, ip=None, start_time=None, end_time=None,
            lifetime=None, secret=None, stream_id=None):
        self._ip = ip
        self._start_time = start_time
        self._end_time = end_time
        self._lifetime = lifetime
        self._secret = secret
        self._stream_id = stream_id

    def generateToken(self):

        #start by checking for Secret
        if self._secret is None or len(self._secret) <= 0:
            raise FastlyTokenError('You must provide a secret.')

        if self._stream_id is None:
            raise FastlyTokenError('You must provide a stream ID.')

        if str(self._start_time).lower() == 'now':
            # Initialize the start time if we are asked for a starting time of
            # now.
            self._start_time = int(time.time())
        elif self._start_time is not None:
            try:
                self._start_time = int(self._start_time)
            except:
                raise FastlyTokenError('start_time must be numeric or \'now\'')

        if self._end_time is not None:
            try:
                self._end_time = int(self._end_time)
            except:
                raise FastlyTokenError('end_time must be numeric.')

        if self._lifetime is not None:
            try:
                self._lifetime = int(self._lifetime)
            except:
                raise FastlyTokenError('lifetime must be numeric.')

        if self._end_time is not None:
            if self._start_time is not None and self._start_time >= self._end_time:
                raise FastlyTokenError('Token start time is equal to or after expiration time.')
        else:
            if self._lifetime is not None:
                if self._start_time is None:
                    self._end_time = time.time() + self._lifetime
                else:
                    self._end_time = self._start_time + self._lifetime
            else:
                raise FastlyTokenError('You must provide an expiration time '
                    '--end_time or a lifetime --lifetime.')

        hash_source = ''
        new_token = ''
        if self._ip is not None:
            new_token += 'ip=%s~' % (self._ip)

        if self._start_time is not None:
            new_token += 'st=%d~' % (self._start_time)

        new_token += 'exp=%d' % (self._end_time)

        hash_source += new_token
        hash_source += '~stream_id=%s' % (self._stream_id)

        token_hmac = hmac.new(
            self._secret,
            hash_source,
            getattr(hashlib, 'sha256')).hexdigest()

        return 'hdnts=%shmac=%s' % (new_token, token_hmac)

if __name__ == '__main__':
    usage = 'python gen_token.py [options]\n'\
            'ie.\n' \
            'python gen_token.py'
    parser = optparse.OptionParser(usage=usage, version=APP_VERSION)
    parser.add_option(
        '-l', '--lifetime',
        action='store', type='string', dest='lifetime',
        help='Token expires after SECONDS. --lifetime or --end_time is mandatory.')
    parser.add_option(
        '-e', '--end_time',
        action='store', type='string', dest='end_time',
        help='Token expiration in Unix Epoch seconds. --end_time overrides --lifetime.')
    parser.add_option(
        '-u', '--stream_id',
        action='store', type='string', dest='stream_id',
        help='STREAMID to validate the token against.')
    parser.add_option(
        '-k', '--key',
        action='store', type='string', dest='secret',
        help='Secret required to generate the token. Do not share this secret.')
    parser.add_option(
        '-s', '--start_time',
        action='store', type='string', dest='start_time',
        help='(Optional) Start time in Unix Epoch seconds. Use \'now\' for the current time.')
    parser.add_option(
        '-i', '--ip',
        action='store', type='string', dest='ip_address',
        help='(Optional) The token is only valid for this IP Address.')
    (options, args) = parser.parse_args()
    try:
        generator = FastlyToken(
            options.ip_address,
            options.start_time,
            options.end_time,
            options.lifetime,
            options.secret,
            options.stream_id)

        new_token = generator.generateToken()
        print('%s' % new_token)
    except FastlyTokenError, ex:
        print('%s\n' % ex)
        sys.exit(1)
    except Exception, ex:
        print(str(ex))
        sys.exit(1)

File: js/gen_token.js
'use strict'
const crypto = require('crypto')
const program = require('commander');

class EdgeAuth {
    constructor(options) {
        this.options = options

        if (!this.options.secret) {
            throw new Error('Secret must be provided to generate a token.')
        }

        if(!this.options.streamId) {
          throw new Error('Stream ID must be provided to generate a token.')
        }

    }

    _generateToken() {
        var startTime = this.options.startTime
        var endTime = this.options.endTime

        if (typeof startTime === 'string' && startTime.toLowerCase() === 'now') {
            startTime = parseInt(Date.now() / 1000)
        } else if (startTime) {
            if (typeof startTime === 'number' && startTime <= 0) {
                throw new Error('startTime must be number ( > 0 ) or "now"')
            }
        }

        if (typeof endTime === 'number' && endTime <= 0) {
            throw new Error('endTime must be number ( > 0 )')
        }

        if (typeof this.options.lifetimeSeconds === 'number' && this.options.lifetimeSeconds <= 0) {
            throw new Error('lifetimeSeconds must be number( > 0 )')
        }

        if (!endTime) {
            if (this.options.lifetimeSeconds) {
                if (!startTime) {
                    startTime = parseInt(Date.now() / 1000)
                }
                endTime = parseInt(startTime) + parseInt(this.options.lifetimeSeconds)
            } else {
                throw new Error('You must provide endTime or lifetimeSeconds')
            }
        }

        if (startTime && (endTime < startTime)) {
            throw new Error('Token will have already expired')
        }

        if (this.options.verbose) {
            console.log("Fastly Token Generation Parameters")

            console.log("    Key/Secret      : " + this.options.secret)
            console.log("    IP              : " + this.options.ip)
            console.log("    Stream ID      : " + this.options.streamId)
            console.log("    Start Time      : " + startTime)
            console.log("    Lifetime(seconds) : " + this.options.lifetimeSeconds)
            console.log("    End Time        : " + endTime)
        }

        var hashSource = []
        var newToken = []

        if (this.options.ip) {
            newToken.push("ip=" + this.options.ip)
        }

        if (this.options.startTime) {
            newToken.push("st=" + startTime)
        }
        newToken.push("exp=" + endTime)

        hashSource = newToken.slice()

        hashSource.push("stream_id=" + this.options.streamId)

        var hmac = crypto.createHmac(
            'sha256',
            this.options.secret
        )

        hmac.update(hashSource.join('~'))
        newToken.push("hmac=" + hmac.digest('hex'))

        return newToken.join('~')
    }

}

module.exports = EdgeAuth
////////////////////////
// END lib/edgeauth.js
////////////////////////

program
	.version('0.1.1')
	.option('-l, --lifetime <n>', 'Token expires after SECONDS. --lifetime or --end_time is mandatory.')
  .option('-e, --end_time <n>', 'Token expiration in Unix Epoch seconds. --end_time overrides --lifetime.')
	.option('-u, --stream_id [value]', 'STREAMID to validate the token against.')
  .option('-k, --key [value]', 'Secret required to generate the token. Do not share this secret.')
  .option('-s, --start_time [value]', "(Optional) Start time in Unix Epoch seconds. Use 'now' for the current time.")
  .option('-i, --ip [value]', '(Optional) The token is only valid for this IP Address.')
	.parse(process.argv);

var ea = new EdgeAuth({
	secret: program.key,
	startTime: program.start_time,
	endTime: program.end_time,
	lifetimeSeconds: program.lifetime,
	ip: program.ip,
	streamId: program.stream_id
})


var token

token = ea._generateToken()

console.log("")
console.log(`hdnts=${token}`)
